ext/gettext/gettext.c

```diff
@@ -191,7 +192,28 @@ PHP_FUNCTION(bindtextdomain)
 	PHP_GETTEXT_DOMAIN_LENGTH_CHECK(1, ZSTR_LEN(domain))
 
 	if (dir == NULL) {
-		RETURN_STRING(bindtextdomain(ZSTR_VAL(domain), NULL));
+		btd_result = bindtextdomain(ZSTR_VAL(domain), NULL);
+		if (btd_result == NULL) {
+			/* It's not POSIX compliant, but musl returns
+			 * NULL rather than an implementation-defined
+			 * default directory when no directory was
+			 * previously bound. Technically there is no
+			 * default directory on musl, but the user can
+			 * provide one (or more) via the MUSL_LOCPATH
+			 * environment variable. For lack of a better
+			 * idea, we return MUSL_LOCPATH if it is set,
+			 * even though it may contain multiple paths.
+			 * We use secure_getenv() because the musl docs
+			 * say that MUSL_LOCPATH will be ignored in the
+			 * scenarios where secure_getenv() will return
+			 * NULL rather than the value of the variable. */
+			musl_locpath = secure_getenv("MUSL_LOCPATH");
```

Comment:
(real) question. Is it more reliable than using `__MUSL__` / completely different ?

Reply:
The `NULL` check actually fixes an (unlikely) segfault on all platforms:

> When bindtextdomain() is called with a non-empty domainname and returns a null pointer, it shall set errno to indicate the error... In addition, the textdomain(), bindtextdomain(), and bind_textdomain_codeset() functions may fail if:
>
> [ENOMEM] Insufficient memory available.

Though I've just extended this to consider `errno` as well. Otherwise, it was going to check `MUSL_LOCPATH` if we ran out of memory.

But once we do get to the `MUSL_LOCPATH` part, I think it's better to avoid a blanket check for `__MUSL__`, because this is a bug that might get fixed. Not only is musl pretty clearly in violation of POSIX, but there's also this line in the implementation:

```C
if (!libc.secure) path = getenv("MUSL_LOCPATH");
/* FIXME: add a default path? */
```

If the bug in musl is ever fixed, we don't want the workaround to be triggered just because `__MUSL__` is true. This way it just keeps working until the older musl is phased out.